djhtml/modes.py

```diff
@@ -1,18 +1,40 @@
+from __future__ import annotations
+
 import re
+from abc import ABC, abstractmethod
+from typing import Any, TypedDict
 
 from .lines import Line
 from .tokens import Token
 
 
-class BaseMode:
+class OffsetDict(TypedDict):
+    relative: int
+    absolute: int
+
+
+class BaseMode(ABC):
     """
     Base class for the different modes.
 
     """
 
+    RAW_TOKENS: list[str]
+    COMMENT_TAGS: list[str]
     MAX_LINE_LENGTH = 10_000
 
-    def __init__(self, source=None, return_mode=None, extra_blocks=None):
+    @abstractmethod
+    def create_token(
+        self, raw_token: str, src: str, line: Line
+    ) -> tuple[Token._Base, "BaseMode"]:
+        ...
```

Comment:
Not sure if an abstract method is the correct solution for the `"BaseMode" has no attribute "create_token"  [attr-defined]` error. Please advise.